### PR TITLE
Remove BinaryFormatter use in PSRP serialization

### PR DESCRIPTION
**Pull request by jborean93**
# PR Summary

Removes the usage of `BinaryFormatter` in PSRP connections.

## PR Context

The `BinaryFormatter` class is marked as obsolete and is planned to be removed in a future .NET version https://github.com/dotnet/designs/blob/main/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md. One location where it is used in pwsh is when serializing the client's time zone. On the server end PowerShell has ignored this property since v3 and even on v2 it has an exception handler to ignore invalid values. This means the client can either omit the property altogether or send an empty value and the server will still be fine.

One minor difference between a missing property and an empty property value is that `$PSSenderInfo.ClientTimeZone` is `$null` when it's missing and set to `[TimeZone]::Current` when it's an empty value. So to preserve as much existing behaviour an empty value was used instead.

The only concern I have with this approach that I haven't tested it with an Exchange endpoint to see if it has any special behaviour with this property.

Partially addressed https://github.com/PowerShell/PowerShell/issues/14054

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by awakecoding** [comment]
Thanks a lot @jborean93 for this pull request, I really appreciate it!

**Comment by PaulHigin** [comment]
I would love to get rid of the `Client Timezone` property.  It is part of the PowerShell remoting protocol (PSRP), but is listed as an optional property.  So it is Ok to not sent it, but **not** Ok to provide an incorrect value.

My main concern is how it might affect Exchange endpoints.  I was not on the team when this property was included, but I seem to recall it was included for Exchange endpoints.  One possible solution could be to serialize this object ourselves 'safely' so that it can be deserialized and used on older endpoints.  But I don't know how important it is and if it is really needed.

Marking for committee review.

File: src/System.Management.Automation/engine/remoting/fanin/PSPrincipal.cs
Comment (PaulHigin) [inline src/System.Management.Automation/engine/remoting/fanin/PSPrincipal.cs:130]: This violates the remoting protocol (PSRP).  We either need to provide the actual client timezone or nothing.

Comment (jborean93) [reply src/System.Management.Automation/engine/remoting/fanin/PSPrincipal.cs:130]: How does it violate the protocol? PowerShell, since v3, has ignored whatever was sent to it from the client and did [TimeZone.CurrentTimeZone](https://docs.microsoft.com/en-us/dotnet/api/system.timezone.currenttimezone?view=net-6.0) as per https://github.com/PowerShell/PowerShell/blob/7cc9c874730e87b06dbe54d50e77c1c472996c54/src/System.Management.Automation/engine/remoting/common/WireDataFormat/EncodeAndDecode.cs#L2376-L2392.

The `TimeZone` class is deprecated and further along PowerShell just changes it to [TimeZoneInfo.Local](https://docs.microsoft.com/en-us/dotnet/api/system.timezoneinfo.local?view=net-6.0)

I can make this `null` and just always return that if you wish, it just seems like having it always set to this value is more aligned with the current behaviour.

**Comment by jborean93** [comment]
> but not Ok to provide an incorrect value.

I can change the code to omit the field altogether if you wish, the only reason why I send an empty byte value is that

+ If the serialization of the TZ fails on the client this is what is sent today
+ the server populates the `$PSSenderInfo.ClientTimeZone` differently based on whether the property was set or not
  + If missing the value is alwasy null
  + If set, v2 will use the provided value and `null` if the value is empty/invalid
  + If set, v3 will use `TimeZone.CurrentTimeZone`

I tested both a missing and empty value and decided on the latter to try and preserve as much existing behaviour as I could (`$PSSenderInfo.ClientTimeZone` not being `null`). Considering currently PowerShell can already send an empty byte array when serialization fails (for whatever reason) it seems like the behaviour is already vaguely defined.

> My main concern is how it might affect Exchange endpoints. I was not on the team when this property was included, but I seem to recall it was included for Exchange endpoints. One possible solution could be to serialize this object ourselves 'safely' so that it can be deserialized and used on older endpoints. But I don't know how important it is and if it is really needed.

Yea I can understand this concern and ultimately was hoping that you, or someone else in the pwsh team, can get them to review this change. I know for sure that Exchange works with the property not present altogether as my Python client doesn't add this field at all, I'm just not sure if it will be able to handle an empty value that this PR has today.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this, we agree that given dotnet's deprecation and removal plan, we should remove our support in 7.3 to get customer feedback on impact.  However, the only viable mitigation we can give them is to use 5.1 or 7.2 instead.

**Comment by daxian-dbw** [comment]
@jborean93 and @PaulHigin, can you please help me understand if additional changes are needed given the committee's comments?

**Comment by PaulHigin** [comment]
The committee Ok'd removing the ClientTimeZone message since it is optional in the protocol.  We want to see if there are any adverse affects.  However, I believe this PR will have to change to not substitute the ClientTimeZone on the server with the current local timezone, as that does not reflect the actual client timezone and violates the protocol.

**Comment by jborean93** [comment]
> I believe this PR will have to change to not substitute the ClientTimeZone on the server with the current local timezone, as that does not reflect the actual client timezone and violates the protocol.

This is currently what happens today, the server (since pwsh v3) will ignore the `ClientTimeZone` value and always use the current time zone value as per https://github.com/PowerShell/PowerShell/blob/7cc9c874730e87b06dbe54d50e77c1c472996c54/src/System.Management.Automation/engine/remoting/common/WireDataFormat/EncodeAndDecode.cs#L2376-L2392.

**Comment by PaulHigin** [comment]
@jborean93 
>Only PowerShell 2 actually read this value and parsed it

How do you know?  Do you have access to the code base?

In any case, this is contrary to what the protocol states.  So either the protocol must change or the code.  This PR should change the code and we can evaluate its effects in preview.

**Comment by jborean93** [comment]
> How do you know? Do you have access to the code base?

The only way I can, through tools like dotpeek and dnsspy. On Windows 7 with pwsh 2 installed the method there will actually deserialize the .NET binary blob. On pwsh v3 and newer (since 2012) it has the same comment you see today in the code.

> In any case, this is contrary to what the protocol states. So either the protocol must change or the code. This PR should change the code and we can evaluate its effects in preview.

That's fine with me, I don't know where you keep the docs, I'm just trying to remove a component that is going to be removed from .NET. My goal is to try and keep the same behaviour as is today.

**Comment by PaulHigin** [comment]
@jborean93
Makes sense, thanks.

It is not just updating protocol docs (which requires approval) but also updating the protocol version since it will be (or has been, since the code is now doing this) changed (ClientTimeZone information is actually ServerTimeZone).  That protocol change makes no sense and is not worth doing in my opinion and as the committee recommends, this PR should simply remove it, since it is optional.  I am thinking the 'ClientTimeZone' property should be set to null.

**Comment by jborean93** [comment]
Considering the change happened in pwsh v3 the protocol version was already bumped from 2.1 to 2.2 I'm not sure it really warrants a new bump for something that has already happened.

I'm happy to just omit the value altogether but it is strictly a breaking change; `$PSSenderInfo.ClientTimeZone` is now `$null` compared to the time zone. I highly doubt people are relying on this property but considering how change adverse this project is I assumed you preferred it to continue to be the server time zone, what it has been since v3. If you still want me to change the PR so it's never sent I can but I just want to make sure you are aware of the potential breaking change it can cause.

**Comment by daxian-dbw** [comment]
Since the property has been set to the server time zone since PowerShell v3, I think we should keep it that way to avoid potentially breaking anything.

**Comment by PaulHigin** [comment]
@daxian-dbw  We cannot keep this behavior, regardless of how long it has been there, as it violates the protocol and we are non-compliant.  We either need to change the protocol (which doesn't make sense) or adhere to the protocol which allows us to not send the message (since it is optional).

**Comment by daxian-dbw** [comment]
OK, so it looks to me that means the server side timezone needs to be fixed regardless of the rest changes in this PR.

**Comment by PaulHigin** [comment]
@daxian-dbw 
No, the server side timezone needs to be fixed as part of this PR.  We cannot remove the TimeZone field from the client session capability message and report an invalid timezone on the server side.

File: src/System.Management.Automation/engine/remoting/fanin/PSPrincipal.cs
Comment (PaulHigin) [reply src/System.Management.Automation/engine/remoting/fanin/PSPrincipal.cs:130]: Yes, please leave it null, as that conforms to the protocol (which makes this capability field optional).  Also I notice that currently this field is null now, with this previous code:

```csharp
#if !CORECLR // TimeZone Not In CoreCLR
                result.TimeZone = TimeZone.CurrentTimeZone;
#endif
```

**Review by PaulHigin — CHANGES_REQUESTED** [review]

**Comment by daxian-dbw** [comment]
@PaulHigin I just pushed a commit to address you comment (leave `ClientTimeZone` to be null). Please review again and see if we can merge this PR. Thanks!

**Review by PaulHigin — APPROVED** [review]

**Comment by jborean93** [comment]
Sorry I've been slack in following it up, don't you also want to remove the line https://github.com/PowerShell/PowerShell/pull/17133/files#diff-24dd8a004527ebae1b7fb5796191281dc116c2f41ed925f4f12b38cab99a71dfR1604 so it doesn't add the property at all in the CLIXML. I believe that was desired rather than an empty byte array.

**Comment by daxian-dbw** [comment]
@PaulHigin Can you comment on @jborean93's message above?

**Comment by PaulHigin** [comment]
Hmm, yes thanks @jborean93.  I think it is best to just not include the TimeZone data item in the message stream.

**Comment by daxian-dbw** [comment]
OK, that line was removed. Can both of you take another look? I will merge if all look good.

**Comment by jborean93** [comment]
The only breaking change I know off is that now is when using a client with this change, the `$PSSenderInfo.ClientTimeZone` value will be `$null` on any sessions created. The same thing would occur if a client then connected to a server running with this change as well. Technically the value before wasn't 100% correct as it was set to the time zone of the server. All in all I don't think it's really a problem IMO.

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by daxian-dbw** [comment]
Thank you both. I will merge this PR today.

**Pull request merged**
